### PR TITLE
Verify access authorization to Websockets endpoints

### PR DESCRIPTION
**Pull request by spena**
### Description 
This is a follow-up PR (part of https://github.com/confluentinc/ksql/pull/2837) that secures Websockets endpoints by checking if a user is authorized to access the endpoint.

It uses a `KsqlAuthorizer`, that should be returned by the `KsqlSecurityExtension`, to verify if the user has access to the KSQL cluster.

The `KsqlDefaultKsqlSecurityExtension` returns a dummy authorizer that allows access to any user. 

### Testing done 
Updated the relevant unit tests and the build passes. Testing the actual functionality of this change requires a non-default implementation of KsqlSecurityExtension.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/security/KsqlSecurityExtension.java [hunk 1]
```diff
@@ -34,21 +34,41 @@
...
+  void register(Configurable<?> configurable, KsqlConfig ksqlConfig);
+
+  /**
+   * Constructs a {@link org.apache.kafka.streams.KafkaClientSupplier} with the user's credentials.
+   *
+   * @param principal The {@link Principal} whose credentials will be used.
+   * @throws KsqlException If an error occurs while creating the
+   * {@link org.apache.kafka.streams.KafkaClientSupplier}.
+   */
+  KafkaClientSupplier getKafkaClientSupplier(Principal principal) throws KsqlException;
```
Comment (vcrfxia) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/security/KsqlSecurityExtension.java]: These two methods shouldn't be here. (They're already in the file below.) Probably the result of a bad merge?

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/security/KsqlSecurityExtension.java [hunk 2]
```diff
@@ -34,21 +34,41 @@
  */
 public interface KsqlSecurityExtension extends AutoCloseable {
   /**
-   * Initializes the security extension. This is called in case the implementation requires
-   * to initializes internal objects prior to registering necessary endpoints.
-   *
-   * @param config The KSQL configuration containing security required configs.
-   * @throws KsqlException If an error occurs while initializing the security extension.
+   * @return The {@code KsqlAuthorizer} used to request access to KSQL resources.
```
Comment (vcrfxia) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/security/KsqlSecurityExtension.java]:
```suggestion
   * @return The {@code KsqlAuthorizer} used to authorize access to KSQL resources.
```

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/security/KsqlSecurityExtension.java [hunk 3]
```diff
@@ -34,21 +34,41 @@
...
 
   /**
-   * Registers the security extension for the KSQL REST endpoints.
+   * Registers the security extension.
+   * </p>
+   * A {@link Configurable} is passed so that the extension can register REST filters to
+   * secure KSQL REST endpoints (i.e. Authorization filters).
    *
    * @param configurable The {@link Configurable} object where to register the security plugins.
+   * @param config The KSQL configuration containing security required configs.
```
Comment (vcrfxia) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/security/KsqlSecurityExtension.java]:
```suggestion
   * @param ksqlConfig The KSQL configuration containing security required configs.
```
or change the name of the param below.

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/security/KsqlAuthorizer.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+ */
+
+package io.confluent.ksql.rest.server.security;
+
+import java.security.Principal;
+
+/**
+ * This Authorizer is used by KSQL to request access to KSQL resources.
+ * </p>
+ * Currently, the authorizer is limited to request access to Websocket endpoints. REST endpoints
```
Comment (vcrfxia) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/security/KsqlAuthorizer.java]:
```suggestion
 * Currently, the authorizer is only used to request access to WebSocket endpoints. REST endpoints
```

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/security/KsqlAuthorizer.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+ */
+public interface KsqlAuthorizer {
+  /**
+   * Checks if a user has access to the specified {@code resourceClass} and {@code resourceMethod}.
+   * </p>
+   * The access requested is simply to check if the {@code user} is authorized to call the desired
+   * REST/Websocket endoiunt or not. It is up to the implementation to decide the type of
+   * access to verify.
+   *
+   * @param principal The user principal requesting access authorization
```
Comment (vcrfxia) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/security/KsqlAuthorizer.java]:
```suggestion
   * @param user The user principal requesting access authorization
```
or rename the param below.

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/WSQueryEndpoint.java [hunk 6]
```diff
@@ -163,10 +164,23 @@ public WSQueryEndpoint(
         Objects.requireNonNull(serviceContextFactory, "serviceContextFactory");
   }
 
+  private void checkEndpointAuthorization(final Principal userPrincipal) {
```
Comment (vcrfxia) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/WSQueryEndpoint.java]: nit: move private method below public methods.

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/WSQueryEndpoint.java [hunk 7]
```diff
@@ -163,10 +164,23 @@ public WSQueryEndpoint(
...
+    final Class className = this.getClass();
+    final String methodName = this.getClass().getAnnotation(ServerEndpoint.class).value();
+
+    if (!securityExtension.getAuthorizer().hasAccess(userPrincipal, className, methodName)) {
+      final String userName = (userPrincipal != null) ? userPrincipal.getName() : null;
+      throw new KsqlException(
+          String.format("User:%s is denied to access this cluster.", userName)
+      );
```
Comment (vcrfxia) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/WSQueryEndpoint.java]: Besides throwing an error we want to close the session with the appropriate error code, right? I see that the unit test you've added checks for the close code `CANNOT_ACCEPT` which makes sense to me. Where is that close code being returned?

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/security/KsqlAuthorizer.java [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+ * </p>
+ * Currently, the authorizer is limited to request access to Websocket endpoints. REST endpoints
+ * should be secured by the implementation when calling {@link KsqlSecurityExtension#register()}.
+ */
+public interface KsqlAuthorizer {
+  /**
+   * Checks if a user has access to the specified {@code resourceClass} and {@code resourceMethod}.
+   * </p>
+   * The access requested is simply to check if the {@code user} is authorized to call the desired
+   * REST/Websocket endoiunt or not. It is up to the implementation to decide the type of
```
Comment (vcrfxia) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/security/KsqlAuthorizer.java]:
```suggestion
   * REST/WebSocket endpoint or not. It is up to the implementation to decide the type of
```

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/WSQueryEndpoint.java [hunk 9]
```diff
@@ -163,10 +164,23 @@ public WSQueryEndpoint(
         Objects.requireNonNull(serviceContextFactory, "serviceContextFactory");
   }
 
+  private void checkEndpointAuthorization(final Principal userPrincipal) {
+    final Class className = this.getClass();
+    final String methodName = this.getClass().getAnnotation(ServerEndpoint.class).value();
```
Comment (rodesai) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/WSQueryEndpoint.java]: Why are we using the path here when we use the method name for the http endpoints?

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/security/KsqlAuthorizer.java [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+   * access to verify.
+   *
+   * @param principal The user principal requesting access authorization
+   * @param resourceClass The REST class name
+   * @param resourceMethod The REST method name
```
Comment (rodesai) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/security/KsqlAuthorizer.java]: this is confusing - when I read this I thought GET/PUT/POST/etc

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/WSQueryEndpoint.java (hunk 6 above)
Comment (spena) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/WSQueryEndpoint.java]: Done

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/WSQueryEndpoint.java (hunk 7 above)
Comment (spena) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/WSQueryEndpoint.java]: The exception is caught un the `inOpen` method. The try/catch that catches it makes sure to close the session:

```
try {
      // Check if the user has authorization to access this Websocket endpoint
      checkEndpointAuthorization(session.getUserPrincipal());
      ....
} catch (final Exception e) {
      log.debug("Error processing request", e);
      SessionUtil.closeSilently(session, CloseCodes.CANNOT_ACCEPT, e.getMessage());
}
```

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/security/KsqlAuthorizer.java (hunk 4 above)
Comment (spena) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/security/KsqlAuthorizer.java]: Done

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/security/KsqlAuthorizer.java (hunk 8 above)
Comment (spena) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/security/KsqlAuthorizer.java]: Done

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/security/KsqlAuthorizer.java (hunk 5 above)
Comment (spena) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/security/KsqlAuthorizer.java]: Done

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/security/KsqlSecurityExtension.java (hunk 2 above)
Comment (spena) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/security/KsqlSecurityExtension.java]: Done

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/security/KsqlSecurityExtension.java (hunk 3 above)
Comment (spena) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/security/KsqlSecurityExtension.java]: Done

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/security/KsqlSecurityExtension.java (hunk 1 above)
Comment (spena) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/security/KsqlSecurityExtension.java]: Done

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/security/KsqlAuthorizer.java (hunk 10 above)
Comment (spena) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/security/KsqlAuthorizer.java]: Got it. I am using REST and Resource class/method names as the same concept. The more I learn about REST, the more I understand I am wrong.

I will change it to `The resource class name` and `The resource method name`.

Comment (spena) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/security/KsqlAuthorizer.java]: Done

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/WSQueryEndpoint.java (hunk 9 above)
Comment (spena) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/WSQueryEndpoint.java]: Done

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/WSQueryEndpoint.java [hunk 11]
```diff
@@ -177,6 +179,9 @@ public void onOpen(final Session session, final EndpointConfig unused) {
       return;
     }
     try {
+      // Check if the user has authorization to access this Websocket endpoint
+      checkEndpointAuthorization(session.getUserPrincipal(), "onOpen");
```
Comment (agavra) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/WSQueryEndpoint.java]: hard-coding `onOpen` is very fragile. does it refer to the method name? the annotation on the method name? some magic constant? If I were a user implementing this, how would I know what to implement? as it stands, I need to examine the code?

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/WSQueryEndpoint.java [hunk 12]
```diff
@@ -241,6 +246,17 @@ public void onError(final Session session, final Throwable t) {
     log.error("websocket error in session {}", session.getId(), t);
   }
 
+  private void checkEndpointAuthorization(final Principal userPrincipal, final String methodName) {
+    final Class className = this.getClass();
+
+    if (!securityExtension.getAuthorizer().hasAccess(userPrincipal, className, methodName)) {
+      final String userName = (userPrincipal != null) ? userPrincipal.getName() : null;
+      throw new KsqlException(
+          String.format("User:%s is denied to access this cluster.", userName)
```
Comment (agavra) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/WSQueryEndpoint.java]: should this be a 403 exception? how does this error surface to the user?  also I think we should have a different error for null user, and we should have a more descriptive error message:
`"Access denied for user %s to %s:%s".format(username, className, methodName)`

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/WSQueryEndpoint.java (hunk 11 above)
Comment (spena) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/WSQueryEndpoint.java]: It's the method name. Any idea how to get it without hard-coding it?

Comment (agavra) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/WSQueryEndpoint.java]: The problem is that perhaps not hard-coding it is even more fragile :D I think a constant would be good enough for now

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/WSQueryEndpoint.java (hunk 12 above)
Comment (spena) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/WSQueryEndpoint.java]: Does the user need to know internal class and method names? For them, they just want to open a session to the KSQL server, and the message clarifies they don't have access to it.

Perhaps when we have finer-grained privileges on KSQL, we could mention what resources are denied and why, and what permissions they require.

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/WSQueryEndpoint.java (hunk 11 above)
Comment (spena) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/WSQueryEndpoint.java]: Done. I used a constant.

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/WSQueryEndpoint.java (hunk 12 above)
Comment (agavra) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/WSQueryEndpoint.java]: I feel pretty strongly that error messages should be actionable. As it stands, I have no idea what I need to do in order to make progress. Showing the resource/method pairing would help me understand what i need to get

**Pull request merged**